keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1154 @@
+heap allocations, by adopting CBOR as a self-describing binary format.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Reduce CPU time and heap churn of encode and decode for cases that can't use
+  Protobuf:
+  - custom resource storage
+  - custom resource serving
+  - dynamic clients
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Replace existing usage of Protobuf.
+- Substantially reduce the size of encoded objects (a modest size reduction is
+  anticipated).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Format
+
+The output of the CBOR encoder is a single tagged data item as specified in
+“[Self-Described
+CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor)”,
+with no additional envelope. Self-described CBOR -- a tagged data item with tag
+number 55799 -- has the same semantics as the same data item with no tag, with
+the convenient property that its encoded form is always prefixed by
+`0xd9d9f7`. By design, this prefix is never found at the beginning of a JSON
+text and can be used as a “magic number” to distinguish the data format of a
+stored object at rest.
+
+To support decoding custom resources that have been stored as a mixture of JSON
+and CBOR, the CBOR serializer will implement RecognizingDecoder by checking for
+the prefix `0xd9d9f7`.
+
+Streaming responses (i.e. watches) will be serialized as CBOR Sequences. A [CBOR
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
+equivalent to the existing JSON stream serialization behavior and takes
+advantage of the property that, like JSON objects – and unlike non-object JSON
+documents, e.g. numbers – CBOR data items are self-delimiting.
+
+### Negotiation
+
+Proactive content negotiation will be supported for clients that want to receive
+CBOR-encoded responses using the MIME type “application/cbor” in the Accept
+request header. For compatibility with API servers that don’t support CBOR,
+clients should also accept “application/json” (with a lower quality factor) and
+choose the appropriate decoder based on the Content-Type response header.
+
+Similarly, streaming requests should use the MIME type for CBOR Sequences,
+“application/cbor-seq”.
+
+CBOR will not be a supported encoding for JSON Patches, JSON Merge Patches,
+Strategic Merge Patches, or Server-Side Apply configurations.
```

Comment:
The first three patch types would have to be transcoded into a JSON document on the server side, undermining the efficiency benefit of CBOR, unless we were to define parallel CBOR versions for each type. Including SSA in this list is probably a mistake. We should have good support for decoding apply configs with CBOR since the proposal already requires good handling of Unstructured for CRs. I will revise this.

Reply:
> The first three patch types would have to be transcoded into a JSON document on the server side

I'm not sure that's true for SMP, but I might be mistaken.

Reply:
I was under the impression that SMP was conceived as an extension of JSON Merge Patch, but this is before my time and I don't have complete context. It's at least described in 2017 as ["a customized version of JSON merge patch"](https://github.com/kubernetes/community/blob/d281f790d1e00260e4237f4503a8cba3a48efa9b/contributors/devel/sig-api-machinery/strategic-merge-patch.md#background).

Reply:
OK, I wasn't sure either so I took a few minutes to actually look into it.

json-patch
-----------
Looks like we decode the bytes directly using github.com/evanphx/json-patch
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L383. Looks like we could possibly do the decoding manually from CBOR, but the validate functions are private, so not entirely obvious. https://github.com/evanphx/json-patch/blob/dc808d1b15ce3e04b8a05e6d31dbd8bcbaf45c76/v5/patch.go#L1079

json-merge-patch
------------------
Looks like the json-merge-patch library takes the json bytes directly. https://github.com/evanphx/json-patch/blob/dc808d1b15ce3e04b8a05e6d31dbd8bcbaf45c76/v5/merge.go#L110. It's a little sad that they just directly unmarshal in a way that we can't replace, so forget about that one.

strategic-merge-patch
-----------------------
This code is entirely internal, and it looks like we unmarshal ourselves into a `map[string]interface{}` leading up to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L694, so I think we can easily replace that if needed.

Reply:
OK yeah, that's fine. I still think we should be able to do the strategic merge patch though.

Reply:
That sounds reasonable. I will make that change -- thanks for bearing with me.